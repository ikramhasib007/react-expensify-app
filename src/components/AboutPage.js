import React from 'react';
import {Link} from 'react-router-dom';

const AboutPage = () => (
  <div>
    <div className="jumbotron jumbotron-fluid p-0">
      <div className="container row m-0 p-0">
        <div className="card col-md-4 p-0 rounded-0 my--5">
          <img
            className="card-img-top rounded-0"
            src="assets/resume/img/profile.jpg"
            alt="Card image cap"
            height= "100%"/>
        </div>
        <div className="col-md-8 p-4">
          <div className="bs-callout bg-white bs-callout-primary mt-0">
            <h5>Software Developer</h5>
            <h6>SCT-Bangla Limited, Banasree, Dhaka.</h6>
            <p className="lead">Uses Technology: PHP, Laravel, AngularJS (1x), Javascript</p>
          </div>
          <div className="d-flex justify-content-around mt-2">
              <a href="https://www.facebook.com/ikramhasib007" target="_blank"><i className="fab fa-facebook fa-3x"></i></a>
              <a href="https://www.linkedin.com/in/ikram-ud-daula/" target="_blank" className="text-info"><i className="fab fa-linkedin fa-3x"></i></a>
              <a href="https://github.com/ikramhasib007" target="_blank" className="text-dark"><i className="fab fa-github-square fa-3x"></i></a>
              <a href="https://stackoverflow.com/users/5461148/ikram-ud-daula" className="text-warning" target="_blank"><i className="fab fa-stack-overflow fa-3x"></i></a>
              <a href="https://www.dropbox.com/s/ucatbh14ymlqplh/IKRAM-UD-DAULA_CV_NEW.pdf?dl=0" title="CV" className="text-secondary" target="_blank"><i className="fas fa-user-circle fa-3x"></i></a>
          </div>
        </div>
      </div>
    </div>
    <div className="bs-callout bg-light bs-callout-success mt-0">
      <h4 className="font-weight-bold">Expensify Project Build By Using</h4>
      <p className="lead text-dark">
        <strong className="pr-1">React.js,</strong>
        <strong className="p-1">Redux,</strong>
        <strong className="p-1">Firebase,</strong>
        <strong className="p-1">ES6</strong>
      </p>
    </div>
  </div>
);

export default AboutPage;