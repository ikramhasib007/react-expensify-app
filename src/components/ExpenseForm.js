import React, {Component} from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import 'react-dates/initialize';

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description :'',
      amount: props.expense ? props.expense.amount.toString() :'',
      note: props.expense ? props.expense.note :'',
      createdAt: props.expense ? moment(props.expense.createdAt) :moment(),
      calendarFocused: false,
      errorDescription: '',
      errorAmount: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description:description, errorDescription:''}));
  }
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({amount:amount, errorAmount:''}));
    }
  }
  onNoteChange = (e) => {
    //   const note = e.target.value;
    e.persist(); // alternative way
    this.setState(() => ({note: e.target.value}));
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}));
    }
  };
  onFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}));
  };
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description) {
      this.setState(() => ({errorDescription: 'Please provide description'}));
    }
    if (!this.state.amount) {
      this.setState(() => ({errorAmount: 'Please provide amount'}));
    } 
    if (this.state.description && this.state.amount) {
      this.props.onSubmit({
          description: this.state.description,
          amount: parseFloat(this.state.amount, 10),
          createdAt: this.state.createdAt.valueOf(),
          note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <label htmlFor="description">Description</label>
            <input
              type="text"
              value={this.state.description}
              onChange={this.onDescriptionChange}
              className={!this.state.errorDescription
              ? 'form-control'
              : 'form-control is-invalid'}
              id="description"
              placeholder="Description"/> {!this.state.errorDescription && <small className="form-text text-muted">Add a description. example: Water bill, Bought a shirt, etc...</small>}
            {this.state.errorDescription && <div className="invalid-feedback">{this.state.errorDescription}</div>}
          </div>
          <div className="row">
            <div className="form-group col">
              <label htmlFor="amount">Amount</label>
              <input
                type="text"
                className={!this.state.errorAmount
                ? 'form-control'
                : 'form-control is-invalid'}
                value={this.state.amount}
                onChange={this.onAmountChange}
                id="amount"
                placeholder="Amount"/> {this.state.errorAmount && <div className="invalid-feedback">{this.state.errorAmount}</div>}
            </div>
            <div className="form-group col">
              <label htmlFor="date">Date</label><br/>
              <SingleDatePicker
                date={this.state.createdAt}
                onDateChange={this.onDateChange}
                focused={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                isOutsideRange={() => false}
                numberOfMonths={1}
                block={true}
                small={true}
                readOnly={true}/>
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="note">Note (optional)</label>
            <textarea
              className="form-control"
              value={this.state.note}
              onChange={this.onNoteChange}
              placeholder="Add a note for your expense (optional)"
              rows="3"></textarea>
          </div>
          <button 
            type="submit" 
            className='btn btn-primary btn-lg btn-block'
          >
            {this.props.expense?'Update Expense':'Add Expense'}
          </button>
        </form>
      </div>
    );
  }
}