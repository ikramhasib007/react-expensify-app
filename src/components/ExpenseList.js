import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    <div className='row mt-4'>
      {
        props.expenses.length === 0 ? (
          <div className="col-sm-6 mb-4"><h4>No expenses</h4></div>
        ) : (
          props.expenses.map((expense) => (
            <ExpenseListItem key={expense.id} {...expense} />
          ))
        )
      }
    </div>
  </div>
);
// console.log(connect);
const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
});
export default connect(mapStateToProps)(ExpenseList);