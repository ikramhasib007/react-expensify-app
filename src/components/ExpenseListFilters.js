import React, {Component} from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';


export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() =>({calendarFocused}));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate(e.target.value);
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount(e.target.value);
    }
  };

  render() {
    return (
      <div className="row">
        <div className="col-lg-4 col-md-6">
          <input
            type="text"
            className="form-control mb-1"
            placeholder="Search..."
            value={this.props.filters.text}
            onChange={this.onTextChange}/>
        </div>
        <div className="col-lg-4 col-md-6">
          <div className="input-group mb-1">
            <div className="input-group-prepend">
              <span className="input-group-text">Sort By</span>
            </div>
            <select
              className="custom-select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
        </div>
        <div className="col-lg-4 col-md-12">
        <DateRangePicker
          startDate={this.props.filters.startDate} 
          startDateId={'fsdfdsdfsfs'}
          endDate={this.props.filters.endDate} 
          endDateId={'fdsfsf'}
          onDatesChange={this.onDatesChange} 
          focusedInput={this.state.calendarFocused} 
          onFocusChange={this.onFocusChange} 
          numberOfMonths={1}
          isOutsideRange={()=>false}
          showClearDates={true}
          block={true}
          readOnly={true}
        />
        </div>
      </div>
    );
  };
};

const mapStateToProps = (state) => {
  return {filters: state.filters};
};
const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (date) => dispatch(setStartDate(date)),
  setEndDate: (date) => dispatch(setEndDate(date))

});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);