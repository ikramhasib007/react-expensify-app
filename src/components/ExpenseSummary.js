import React from 'react';
import {connect} from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpenseSummary = (props) => {
    return <h1>Viewing {props.expenseCount.length} expense{props.expenseCount.length>1 && 's'} totalling {numeral(props.expensesTotal / 100).format('$0,0.00')}</h1>
}

const mapStateToProps = (state) => ({
    expenseCount: getVisibleExpenses(state.expenses, state.filters),
    expensesTotal: getExpensesTotal(getVisibleExpenses(state.expenses, state.filters))
});
 
export default connect(mapStateToProps)(ExpenseSummary);