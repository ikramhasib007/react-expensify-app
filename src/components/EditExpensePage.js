import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startEditExpense, startRemoveExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/dashboard');
  }

  render() {
    return (
      <div className="row">
        <div className="col-md-6">
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        </div>
        <div className="col-md-2 offset-md-2 mt-4">
        <button
          className="btn btn-outline-danger btn-lg btn-block"
          onClick={() => {
              this.props.startRemoveExpense({id:this.props.expense.id});
              this.props.history.push('/dashboard');
          }}
        >Remove</button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id,expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);