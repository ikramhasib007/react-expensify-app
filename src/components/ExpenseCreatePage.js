import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {startAddExpense} from '../actions/expenses';

export class ExpenseCreatePage extends React.Component {
  onSubmit = (expense) => {
    //props.dispatch(startAddExpense(expense));
    this.props.startAddExpense(expense);
    this.props.history.push('/dashboard');
  }
  render () {
    return (
      <div className="row">
        <div className="col-md-6">
          <ExpenseForm
            onSubmit={this.onSubmit}/>
        </div>
      </div>
    );
  }
};

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(ExpenseCreatePage);