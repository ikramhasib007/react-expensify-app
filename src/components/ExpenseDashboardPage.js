import React from 'react';
import ExpenseList from './ExpenseList';
import Container from '../hoc/Auxiliary';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);


export default ExpenseDashboardPage;