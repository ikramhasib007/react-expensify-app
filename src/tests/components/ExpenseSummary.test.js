import React from 'react';
import {ExpenseSummary} from '../../components/ExpenseSummary';
import {shallow} from 'enzyme';

test("should render ExpenseSummary correctly with 1 expense", () => {
    const wrapper = shallow(<ExpenseSummary expensesTotal={123} expenseCount={1} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary correctly with multiple expenses", () => {
    const wrapper = shallow(<ExpenseSummary expensesTotal={34234} expenseCount={25} />);
    expect(wrapper).toMatchSnapshot();
});
