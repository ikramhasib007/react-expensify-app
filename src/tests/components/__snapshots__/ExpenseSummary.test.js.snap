// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly with 1 expense 1`] = `
<h1>
  Viewing 
   expense
   totalling 
  $1.23
</h1>
`;

exports[`should render ExpenseSummary correctly with multiple expenses 1`] = `
<h1>
  Viewing 
   expense
   totalling 
  $342.34
</h1>
`;
